### ISSUE TITLE
[Bug]: Router does not immediately fallback when Groq is rate limited

### ISSUE THREAD

**Issue by AshSourceTable** [BerriAI/litellm#3274]
### What happened?

When Groq returns a 429 Too Many Requests error, rather than immediately switching to the fallback, the router is retrying 2 times after a 47-second wait, which is not what I have configured:

```
[INFO] HTTP Request: POST https://api.groq.com/openai/v1/chat/completions "HTTP/1.1 429 Too Many Requests"
[INFO] Retrying request to /chat/completions in 47 seconds
```
...the above log lines repeat twice more.

We then finally get a proper exception and it falls back as it should:
```
LiteLLM Router:INFO: router.py:409 - litellm.completion(model=groq/mixtral-8x7b-32768) Exception GroqException - Error code: 429 - {'error': {'message': 'Rate limit reached for model `mixtral-8x7b-32768` in organization `xxx` on tokens per minute (TPM): Limit 9000, Used 0, Requested ~16008. Please try again in 46.72s.', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
```
...the above log line repeats once more.

Here's the code I'm running. Note allowed_fails and num_retries = 0:
```python
# Generate a long_string that will trigger the TPM error
while len(long_string.encode('utf-8')) < 80000:
    long_string += "hello "
messages = [{"role": "user", "content": long_string}]
model_list = [
    {'model_name': 'groq/mixtral-8x7b-32768', 'litellm_params': {'model': 'groq/mixtral-8x7b-32768', 'timeout': 30, 'stream_timeout': 5}}, 
    {'model_name': 'gpt-4-turbo', 'litellm_params': {'model': 'gpt-4-turbo', 'timeout': 30, 'stream_timeout': 5}}]
router = litellm.Router(
    model_list=model_list,
    allowed_fails=0,
    cooldown_time=120,
    num_retries=0,
    retry_after=1,
    fallbacks=[{'groq/mixtral-8x7b-32768': ['gpt-4-turbo']}],
    context_window_fallbacks=[{'groq/mixtral-8x7b-32768': ['gpt-4-turbo']}],
    enable_pre_call_checks=True
)
return router.completion('groq/mixtral-8x7b-32768', messages=messages, stream=True)
```

**Comment by ghost** [comment]
hey @AshSourceTable acknowledging this. testing now

**Comment by ghost** [comment]
Hey @AshSourceTable which version is this on? 

I tested it on colab + local, and it shouldn't even be making the call to groq since you have enable_pre_call_check

**without fallbacks**
this is the error i see
```
...
    raise e
  File "/opt/homebrew/lib/python3.11/site-packages/litellm/router.py", line 359, in _completion
    deployment = self.get_available_deployment(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/litellm/router.py", line 2965, in get_available_deployment
    healthy_deployments = self._pre_call_checks(
                          ^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/lib/python3.11/site-packages/litellm/router.py", line 2730, in _pre_call_checks
    raise litellm.ContextWindowExceededError(
litellm.exceptions.ContextWindowExceededError: Context Window exceeded for given call
```

**With fallbacks**
```
chunk: ModelResponse(id='chatcmpl-9HkcF1TJEJ4jHLtjSJu7W6J5folvp', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='Hello', role='assistant', function_call=None, tool_calls=None), logprobs=None)], created=1714017801, model='gpt-4-turbo', object='chat.completion.chunk', system_fingerprint='fp_76f018034d', usage=Usage())
chunk: ModelResponse(id='chatcmpl-9HkcF1TJEJ4jHLtjSJu7W6J5folvp', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='!', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1714017802, model='gpt-4-turbo', object='chat.completion.chunk', system_fingerprint='fp_76f018034d', usage=Usage())
chunk: ModelResponse(id='chatcmpl-9HkcF1TJEJ4jHLtjSJu7W6J5folvp', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' How', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1714017802, model='gpt-4-turbo', object='chat.completion.chunk', system_fingerprint='fp_76f018034d', usage=Usage())
chunk: ModelResponse(id='chatcmpl-9HkcF1TJEJ4jHLtjSJu7W6J5folvp', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' can', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1714017802, model='gpt-4-turbo', object='chat.completion.chunk', system_fingerprint='fp_76f018034d', usage=Usage())
chunk: ModelResponse(id='chatcmpl-9HkcF1TJEJ4jHLtjSJu7W6J5folvp', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' I', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1714017802, model='gpt-4-turbo', object='chat.completion.chunk', system_fingerprint='fp_76f018034d', usage=Usage())
chunk: ModelResponse(id='chatcmpl-9HkcF1TJEJ4jHLtjSJu7W6J5folvp', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' assist', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1714017802, model='gpt-4-turbo', object='chat.completion.chunk', system_fingerprint='fp_76f018034d', usage=Usage())
chunk: ModelResponse(id='chatcmpl-9HkcF1TJEJ4jHLtjSJu7W6J5folvp', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' you', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1714017802, model='gpt-4-turbo', object='chat.completion.chunk', system_fingerprint='fp_76f018034d', usage=Usage())
chunk: ModelResponse(id='chatcmpl-9HkcF1TJEJ4jHLtjSJu7W6J5folvp', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content=' today', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1714017802, model='gpt-4-turbo', object='chat.completion.chunk', system_fingerprint='fp_76f018034d', usage=Usage())
chunk: ModelResponse(id='chatcmpl-9HkcF1TJEJ4jHLtjSJu7W6J5folvp', choices=[StreamingChoices(finish_reason=None, index=0, delta=Delta(content='?', role=None, function_call=None, tool_calls=None), logprobs=None)], created=1714017802, model='gpt-4-turbo', object='chat.completion.chunk', system_fingerprint='fp_76f018034d', usage=Usage())
chunk: ModelResponse(id='chatcmpl-9HkcF1TJEJ4jHLtjSJu7W6J5folvp', choices=[StreamingChoices(finish_reason='stop', index=0, delta=Delta(content=None, role=None, function_call=None, tool_calls=None), logprobs=None)], created=1714017802, model='gpt-4-turbo', object='chat.completion.chunk', system_fingerprint='fp_76f018034d', usage=Usage())
```

**Comment by ghost** [comment]
Without `enable_pre_call_check: true` i'm able to see the 3 repeat request issue. working on it now.

**Comment by ghost** [comment]
this is caused b/c of the openai client default of 2 retries. I had an initial fix for it but hit a couple issues. Hoping to have the override working by tomorrow.

**Comment by AshSourceTable** [comment]
@krrishdholakia thanks very much for jumping on this! I've tested some more, using the latest release 1.35.26. Not sure if your fix is in that release but I do see relevant code changes. I found that this issue arises in a specific case: when the context window is _not_ too large for the given model, but the tokens _do_ exceed the TPM, which for Groq is currently 9000.

In that case, the pre-check is passed, the query is POSTed, a 429 error is returned, and two retries are made, despite `retries=0`.

Here's the logs, from after the initial POST:
```
[httpx,_send_single_request:1729] [INFO] HTTP Request: POST https://api.groq.com/openai/v1/chat/completions "HTTP/1.1 429 Too Many Requests"
[openai._base_client,_retry_request:1603] [INFO] Retrying request to /chat/completions in 0.886601 seconds
[httpx,_send_single_request:1729] [INFO] HTTP Request: POST https://api.groq.com/openai/v1/chat/completions "HTTP/1.1 429 Too Many Requests"
[openai._base_client,_retry_request:1603] [INFO] Retrying request to /chat/completions in 1.631223 seconds
[httpx,_send_single_request:1729] [INFO] HTTP Request: POST https://api.groq.com/openai/v1/chat/completions "HTTP/1.1 429 Too Many Requests"

LiteLLM Router:INFO: router.py:537 - litellm.acompletion(model=groq/mixtral-8x7b-32768) Exception GroqException - Error code: 429 - {'error': {'message': 'Rate limit reached for model `mixtral-8x7b-32768` in organization `xxx` on tokens per minute (TPM): Limit 9000, Used 0, Requested ~20023. Please try again in 1m13.486666666s. Visit https://console.groq.com/docs/rate-limits for more information.', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
[LiteLLM Router,_acompletion:537] [INFO] litellm.acompletion(model=groq/mixtral-8x7b-32768) Exception GroqException - Error code: 429 - {'error': {'message': 'Rate limit reached for model `mixtral-8x7b-32768` in organization `xxx` on tokens per minute (TPM): Limit 9000, Used 0, Requested ~20023. Please try again in 1m13.486666666s. Visit https://console.groq.com/docs/rate-limits for more information.', 'type': 'tokens', 'code': 'rate_limit_exceeded'}}
```

...it then proceeds to the fallover model.

**Comment by ghost** [comment]
@AshSourceTable can you please share a colab with the snippet behaving as you show on logs. 

I don't see this happening **with** `enable_pre_call_checks: true`

**Comment by AshSourceTable** [comment]
@krrishdholakia the colab is here:
https://colab.research.google.com/drive/1RvGGg0-AdsxPEmVwJBOf_Pc-8OLgAwTU?usp=sharing

You are absolutely right. The colab does not behave the same as when running locally. The code is absolutely identical, and I am using litellm 1.35.26.

I cannot imagine what's happening here. But it looks like you've already pushed a fix, which is epic! Testing it now.

**Comment by AshSourceTable** [comment]
@krrishdholakia tested commit 160acc and it seems to be working great locally now! Yay!

**Comment by ghost** [comment]
nice. @AshSourceTable curious, would you want a LiteLLM-hosted proxy to do the loadbalancing/fallback logic? 

- What would a good solution here look like for you?

**Comment by AshSourceTable** [comment]
@krrishdholakia maaaaybe? Depending on how it's set up, that has the potential to become a different single-source-of-failure. Also some concern about any impact on latency from the additional hop. But TBH I haven't looked hard at the LiteLLM proxy yet to see what it gives us that the base library + redis doesn't. 

That said I think it's a solid business proposition. Even if it's not for us, I'm sure many folks who don't already have their infrastructure built out will find value in it.

**Comment by ghost** [comment]
closing issue as this is now fixed in v`1.35.29`
